src/index.js

```diff
@@ -266,64 +314,129 @@ class CopyPlugin {
           compilation.fileDependencies.add(file.absoluteFrom);
         }
 
-        logger.debug(`reading "${file.absoluteFrom}" to write to assets`);
+        let itemCache;
+        let source;
 
-        let data;
+        // TODO logger
+        if (cache) {
+          let snapshot;
 
-        try {
-          data = await readFile(inputFileSystem, file.absoluteFrom);
-        } catch (error) {
-          compilation.errors.push(error);
+          try {
+            snapshot = await CopyPlugin.createSnapshot(
+              compilation,
+              file.absoluteFrom
+            );
+          } catch (error) {
+            compilation.errors.push(error);
 
-          return;
-        }
+            return;
+          }
+
+          if (snapshot) {
+            let isValidSnapshot;
 
-        if (pattern.transform) {
-          logger.log(`transforming content for "${file.absoluteFrom}"`);
-
-          if (pattern.cacheTransform) {
-            const cacheDirectory = pattern.cacheTransform.directory
-              ? pattern.cacheTransform.directory
-              : typeof pattern.cacheTransform === 'string'
-              ? pattern.cacheTransform
-              : findCacheDir({ name: 'copy-webpack-plugin' }) || os.tmpdir();
-            let defaultCacheKeys = {
-              version,
-              transform: pattern.transform,
-              contentHash: crypto.createHash('md4').update(data).digest('hex'),
-            };
-
-            if (typeof pattern.cacheTransform.keys === 'function') {
-              defaultCacheKeys = await pattern.cacheTransform.keys(
-                defaultCacheKeys,
-                file.absoluteFrom
+            try {
+              isValidSnapshot = await CopyPlugin.checkSnapshotValid(
+                compilation,
+                snapshot
               );
-            } else {
-              defaultCacheKeys = {
-                ...defaultCacheKeys,
-                ...pattern.cacheTransform.keys,
-              };
+            } catch (error) {
+              compilation.errors.push(error);
+
+              return;
             }
 
-            const cacheKeys = serialize(defaultCacheKeys);
+            itemCache = cache.getItemCache(file.relativeFrom, null);
 
-            try {
-              const result = await cacache.get(cacheDirectory, cacheKeys);
+            if (isValidSnapshot) {
+              try {
+                source = await itemCache.getPromise();
+              } catch (error) {
+                compilation.errors.push(error);
 
-              logger.debug(
-                `getting cached transformation for "${file.absoluteFrom}"`
-              );
+                return;
+              }
+            }
+          }
+        }
```

Comment:
`checkSnapshotValid` on a fresh created snapshot will always be true. You need to call `checkSnapshotValid` on a snapshot that was created in the last run of webpack. The snapshot need to be stored and restored in and from the cache.

Here is some pseudo code:

``` js
// 1. check cache first, validate snapshot in cache entry
cacheEntry = await cache.getPromise(file.relativeFrom, null);
if (cacheEntry !== undefined) {
  const { snapshot, source } = cacheEntry;
  if (await checkSnapshotValid(snapshot)) {
    // return early when cache entry can be used
    return source;
  }
}
// 2. create a new snapshot
const snapshot = await createSnapshot(file.relativeFrom);
// 3. create new data
const source = await createNewSource(file.relativeFrom);
// 4. store both in cache
await cache.storePromise(file.relativeFrom, null, { source, snapshot });
// 5. return new data
return source;
```

You are in the good position that you can call 2. before 3. Usually that's not possible, because dependencies are not known before execution. In that case one would do this:

``` js
// 2a.
const startTime = Date.now();
// 2b. create new data
const source = await createNewSource(file.relativeFrom);
// 3. create a new snapshot
const snapshot = await createSnapshot(file.relativeFrom, startTime);
```